### ISSUE TITLE
Indirect associations

### ISSUE THREAD

**Issue by overlookmotel** [sequelize/sequelize#3935]
Sometimes models have relationships between them which rely on intermediate models.

Here's an example:

Data about users is split into two models `User` and `Profile` (with `User` containing core authentication info, and `Profile` containing more general public info about the user) and then a `Profile` can have several `Email` records associated with it.

```js
var User = sequelize.define('user', {
    username: Sequelize.STRING(50),
    password: Sequelize.STRING(50)
} );
var Profile = sequelize.define('profile', {
    firstName: Sequelize.STRING(50),
    lastName: Sequelize.STRING(50)
} );
var Email = sequelize.define('email', {
    email: {
        type: Sequelize.STRING(200),
        unique: true
    }
} );

User.hasOne(Profile);
Profile.belongsTo(User);

Profile.hasMany(Email);
Email.belongsTo(Profile);
```

Now, if you want to get all the `Email`s for a `User`, you have to go through `Profile`:

```js
User.find( {
    where: {id: 123},
    include: {
        model: Profile,
        include: Email
    }
} ).then( function(user) {
    var emails = user.profile.emails;
} );
```

It would be really useful if you could define the indirect association on the model with something like:

```js
User.hasMany(Email, {
    through: {
        model: Profile,
        include: Email
    }
} );
Email.belongsTo(User, {
    through: {
        model: Profile,
        include: User
    }
} );
```

And then just do:

```js
User.find( {
    where: {id: 123},
    include: Email
} ).then( function(user) {
    var emails = user.emails;
    // do something with emails
} );
```

or:

```js
user.getEmails().then( function(emails) {
    // do something with emails
} );
```

This should also work for very complex associations through multiple join tables.

The intention is to:
* Define relationships between models in the model definitions rather than in code which fetches those associations -  which in my opinion is where they belong
* Remove boiler-plate code every time you want to `find()` that association
* Therefore make code more maintainable as you can switch around the underlying implementation of an association (e.g. changing the `Profile` to `Email` relationship to many-to-many) without having to change the code everywhere you fetch that association

My use case is that I'm building apps with a lot of complex nested relationships between models, and constantly find myself writing the same `find({ include: ... })` code over and over again. It would be great to be able to cut that down.

Does anyone have any thoughts on this?

**Comment by mickhansen** [comment]
That's actually a feature request that pops up now and again (some even expecting non-N:M relations to support through).

However i might suggest using an association reference as the through argument:

```js
User.hasOne(Profile);
Profile.belongsTo(User);

Profile.hasMany(Email);
Email.Profile = Email.belongsTo(Profile);
```

```js
Email.belongsTo(User, {
  through: Email.Profile
});
```

But i suppose we could trivially support both :)

**Comment by overlookmotel** [comment]
@mickhansen Do you think this would be hard to implement? I don't know the innards of how `through` is dealt with inside Sequelize.

Yes, supporting associations as the value of `through` could be very nice, as well as an `include` syntax.

**Comment by mickhansen** [comment]
We only support `through` for N:M.
through for 1:1 or 1:M would work differently - I don't think it would be too tricky though.

**Comment by overlookmotel** [comment]
I was thinking it would be great if it also supported nested relationships.

Like: A `Tour` has many `Concert`s, a `Concert` has many `Band`s, a `Band` has many `Musician`s.

```js
Tour.hasMany(Concert);
Concert.hasMany(Band);
Band.hasMany(Musician);

Tour.hasMany(Musician, {
    through: {
        model: Concert,
        include: {
            model: Band,
            include: Musician
        }
    }
} );
```

Then you could do:

```js
Tour.getMusicians().then( function (musicians) {
    // email musicians the tour schedule
} );
```

This kind of case is where the feature would really come into it's own, as it would cut out a huge amount of `model.find( { include: ... } )` logic.

Would that make it much more difficult to implement?

**Comment by overlookmotel** [comment]
Or using associations, as you suggested:

```js
Tour.hasMany(Concert);
Concert.hasMany(Band);
Band.hasMany(Musician);

Tour.hasMany(Band, { through: Tour.associations.concerts } );
Tour.hasMany(Musician, { through: Tour.associations.bands } );
```

**Comment by mickhansen** [comment]
Hmm, i suppose you'd just need some recursive logic to build the join. However the syntax for defining a nested through model is not that clear imo.

**Comment by overlookmotel** [comment]
@mickhansen Do you have another idea for the nested through model syntax?

Maybe it'd be best only to support the association syntax, to keep it simple.

```js
model.hasMany( anotherModel, { through: association } )
```

**Comment by overlookmotel** [comment]
Related issue: https://github.com/sequelize/sequelize/issues/3845

**Comment by overlookmotel** [comment]
I'm still keen on implementing this at some point.

**Comment by JoshuaWise** [comment]
Is this issue being looked at? I'm trying to implement an indirect relationship but sequelize won't let me express my database design.

**Comment by mikefowler** [comment]
@mickhansen has any further thought been given to this? It seems that this is a pretty common type of relationship that those coming from Rails-land will expect to be possible.

**Comment by andrew-st-angelo-77media** [comment]
I agree that this would be very useful.  I use the same indirect/pass-through association in Rails, frequently.  Further, if what you're interested is data associated with the secondary model, if would save bandwidth on a query to not have to pull down the intermediate model, as part of a `Primary.find({ ... , include: { model: NonTrivialIntermediate, include: NeededModels }})`

**Comment by ph55** [comment]
> opened this issue on Jun 15, 2015

Feels bad man.
